### PR TITLE
Actuator: bug fix - added cycle for CombinedFixedPositionActuator handles now positions close to e.g. 2pi and thus close to zero

### PR DESCRIPTION
**Pull request by Sabrina85**

### CODE REVIEW

File: src/odemis/driver/actuator.py [hunk 1]
```diff
@@ -1404,7 +1404,8 @@ def __init__(self, name, role, children, caxes_map, axis_name, positions, fallba
                         positions. Fallback position can be equal to one of the positions allowed. If fallback is not
                         equal to one of the positions allowed, it is not possible to request moving to this fallback
                         position.
-        atol: absolute tolerance in the position of each child
+        atol (float, float): absolute tolerance in the position of each child
+        cycle (float, float): one complete turn of rotators until they reach position 0 again
```
Comment (pieleric) [inline src/odemis/driver/actuator.py]: `float or None`, where None means "infinity" = no modulo.

File: src/odemis/driver/actuator.py [hunk 2]
```diff
@@ -1479,9 +1481,13 @@ def _readPositionfromChildren(self):
...
 
-        # check whether children positions are in consistency with positions allowed
+        # check whether current children positions are in consistency with positions allowed
         for pos_key, pos_children in self._positions.items():
-            for ap, cp, atol in zip(pos_children_cur, pos_children, self._atol):
+            for ap, cp, atol, cyl in zip(pos_children_cur, pos_children, self._atol, self._cycle):
+                # handle positions close to cycle and thus also close to zero
+                # and within tolerance so util.almost_equal compare the correct values
+                if cyl - ap < atol:
+                    ap = cyl - ap
```
Comment (pieleric) [inline src/odemis/driver/actuator.py]: I don't think it handles correctly the reverse case where cp is almost as large as cyl (eg, 6.2) and ap is > 0.
I think that you could write (but you need to check):
```
if cyl is None:
    dist = abs(ap - cp)
else:
    dist = min((ap - cp) % cyl, (cp - ap) % cyl)

if dist > atol:
    break
```

File: src/odemis/driver/test/actuator_test.py [hunk 3]
```diff
@@ -742,7 +742,8 @@ def setUp(self):
         self.axis2 = "linear"
         self.axis_name = "pol"
         self.atol = [3.392e-5, 3.392e-5]
-        self.fallback = "unspecified position"
+        self.cycle = [math.pi * 2, math.pi * 2]
```
Comment (pieleric) [inline src/odemis/driver/test/actuator_test.py]: Please, also test without cycle. (here would be fine).

File: src/odemis/driver/actuator.py [hunk 4]
```diff
@@ -1423,6 +1424,7 @@ def __init__(self, name, role, children, caxes_map, axis_name, positions, fallba
         self._axis_name = axis_name
         self._positions = positions
         self._atol = atol
+        self._cycle = cycle
```
[the remark this answers is not in the document]
Comment (Sabrina85) [reply src/odemis/driver/actuator.py]: Is it hypothetically possible that cycle is different for the two actuators?

[the remark this answers is not in the document]
Comment (Sabrina85) [reply src/odemis/driver/actuator.py]: Why check that cycle is of len(children) if we multiply it anyways with len(children)?

File: src/odemis/driver/actuator.py [hunk 5]
```diff
@@ -1404,7 +1404,9 @@ def __init__(self, name, role, children, caxes_map, axis_name, positions, fallba
                         positions. Fallback position can be equal to one of the positions allowed. If fallback is not
                         equal to one of the positions allowed, it is not possible to request moving to this fallback
                         position.
-        atol: absolute tolerance in the position of each child
+        atol (float, float): absolute tolerance in the position of each child
+        cycle (float or None): one complete turn of rotators until they reach position 0 again;
```
Comment (pieleric) [inline src/odemis/driver/actuator.py]: `cycle (list of (float or None)): for each axis, the length of a full rotation until it reaches position 0 again`...

File: src/odemis/driver/actuator.py [hunk 6]
```diff
@@ -1419,11 +1421,23 @@ def __init__(self, name, role, children, caxes_map, axis_name, positions, fallba
             if not (len(pos) == 2 or type(pos) == str):
                 raise ValueError("Position %s needs to be of format list with exactly two entries. "
                                  "Got instead position %s." % (key, pos))
+        # TODO the default len for atol is 2
```
Comment (pieleric) [inline src/odemis/driver/actuator.py]: Indeed, you probably could also change it to `None`, and replace that on the fly to `(0,) * len(children)`.

File: src/odemis/driver/test/actuator_test.py [hunk 7]
```diff
@@ -908,11 +910,70 @@ def test_reference(self):
...
+                                                 caxes_map=[self.axis1, self.axis2],
+                                                 positions=self.positions,
+                                                 atol=self.atol,
+                                                 cycle=self.cycle,
+                                                 fallback=self.fallback)
+
+    def test_moveAbs(self):
+        """test position close to 2pi and therefore
+        also close to zero"""
+
```
Comment (pieleric) [inline src/odemis/driver/test/actuator_test.py]: Please, start the test by moving the pol-analyser to a position _not_ `horizontal`.

File: src/odemis/driver/actuator.py [hunk 8]
```diff
@@ -1477,12 +1491,22 @@ def _readPositionfromChildren(self):
...
         """
+        # get current children positions
         pos_children_cur = [c.position.value[ca] for c, ca in zip(self._children, self._caxes_map)]
 
-        # check whether children positions are in consistency with positions allowed
+        # check whether current children positions are in consistency with children positions allowed
         for pos_key, pos_children in self._positions.items():
-            for ap, cp, atol in zip(pos_children_cur, pos_children, self._atol):
-                if isinstance(cp, numbers.Real) and not util.almost_equal(ap, cp, atol=atol, rtol=0):
+            # ap: current children position, cp: children position possible
```
Comment (pieleric) [inline src/odemis/driver/actuator.py]: How about renaming the variables to `cp` = current child position, and `tp` = target child position (or `ep` = expected child position)?

File: src/odemis/driver/actuator.py [hunk 9]
```diff
@@ -1477,12 +1491,22 @@ def _readPositionfromChildren(self):
...
+            for ap, cp, atol, cyl in zip(pos_children_cur, pos_children, self._atol, self._cycle):
+                # handle positions close to cycle and thus also close to zero
+                # and within tolerance so util.almost_equal compare the correct values
+                if cyl is None:
+                    dist = abs(ap - cp)
+                else:
+                    dist = min((ap - cp) % cyl, (cp - ap) % cyl)
+                # if the first pos matches, check also the next. Else don't check any further
```
Comment (pieleric) [inline src/odemis/driver/actuator.py]: That's already explained in the comment next to the `else:`. No need to repeat it here.

File: src/odemis/driver/actuator.py [hunk 10]
```diff
@@ -1434,10 +1430,15 @@ def __init__(self, name, role, children, caxes_map, axis_name, positions, fallba
 
         if cycle is None:
             self._cycle = (None,) * len(children)
+        if atol is None:
```
Comment (pieleric) [inline src/odemis/driver/actuator.py]: Yes, but then change the default value of `atol` in the arguments.

**Pull request merged**